### PR TITLE
Add support for options or amount in refund

### PR DESCRIPTION
**Pull request by msjohns1**

### CODE REVIEW

File: lib/Net/Braintree/Transaction.pm [hunk 1]
```diff
@@ -50,9 +51,12 @@ sub void {
 }
 
 sub refund {
+  # Follow-on credit
   my ($class, $id, $amount) = @_;
-  my $params = {};
-  $params->{'amount'} = $amount if $amount;
+  my $params = ref ($amount) eq 'HASH'
```
Comment (EvanHahn) [inline lib/Net/Braintree/Transaction.pm]: Does this behave differently when `$amount` is falsy (for example, when it's `0`)? We don't want to introduce a breaking change.

File: lib/Net/Braintree/TransactionGateway.pm [hunk 2]
```diff
@@ -42,6 +42,7 @@ sub void {
 
 sub refund {
   my ($self, $id, $params) = @_;
+  confess 'ArgumentError' unless verify_params($params, refund_signature);
```
Comment (EvanHahn) [inline lib/Net/Braintree/TransactionGateway.pm]: We use double quotes in many other places—could we change this to be double quotes?

File: t/transaction.t [hunk 3]
```diff
@@ -7,6 +7,15 @@ subtest "params tests" => sub {
   should_throw("ArgumentError", sub { Net::Braintree::Transaction->sale({"foo" => "Bar"}); });
 };
 
+subtest "refund params test" => sub {
```
Comment (EvanHahn) [inline t/transaction.t]: Could we add more thorough tests here? For example, can we test passing different hashes into `refund`?

**Comment by msjohns1** [comment]
On 03/08/2017 03:44 PM, Evan Hahn wrote:
> EvanHahn requested changes on this pull request.
> 
> This looks good, but we've got a couple of questions. Please pardon our lack of Perl expertise.
> 
>>    my ($class, $id, $amount) = @_;
> -  my $params = {};
> -  $params->{'amount'} = $amount if $amount;
> +  my $params = ref ($amount) eq 'HASH'
> 
> Does this behave differently when `$amount` is falsy (for example, when it's `0`)? We don't want to introduce a breaking change.

The full diff shows a bit more context:

   my ($class, $id, $amount) = @_;
-  my $params = {};
-  $params->{'amount'} = $amount if $amount;
+  my $params = ref ($amount) eq 'HASH'
+    ? $amount
+    : { amount => $amount, }
+  ;

I was using the Ruby library as a guide to be sure the behavior was the
same regardless of which library is used. The Ruby code this change is
supposed to mirror is:

>     def refund(transaction_id, amount_or_options = nil)
>       options = if amount_or_options.is_a?(Hash)
>                   amount_or_options
>                 else
>                   { :amount => amount_or_options }
>                 end

So if I misread that, and having a false or 0 value for
amount_or_options is behaving differently than a false value in the
context of the Perl changes, please let me know and I will match
accordingly.

>> @@ -42,6 +42,7 @@ sub void {
>  
>  sub refund {
>    my ($self, $id, $params) = @_;
> +  confess 'ArgumentError' unless verify_params($params, refund_signature);
> 
> We use double quotes in many other placesâ€”could we change this to be double quotes?

In Perl, double quotes are variable interpolated, and single are not.
But since the pattern is established to double quote regardless, I'll
change it.

>> @@ -7,6 +7,15 @@ subtest "params tests" => sub {
>    should_throw("ArgumentError", sub { Net::Braintree::Transaction->sale({"foo" => "Bar"}); });
>  };
>  
> +subtest "refund params test" => sub {
> 
> Could we add more thorough tests here? For example, can we test passing different hashes into `refund`?

You want more tests that fail the refund_signature() validation routine?
I was following the pattern of tests already in place that caught a
single instance of an invalid argument construction for an ArgumentError
condition in Net::Braintree::Transaction.

**Comment by EvanHahn** [comment]
@msjohns1 Per [my comment in #18](https://github.com/braintree/braintree_perl/pull/18#issuecomment-293314798), please feel free to reopen this pull request on [the community fork](https://github.com/singingfish/braintree_perl). Thanks!

**Pull request closed without merging**